pkg/controller/operators/catalog/querier.go

```diff
@@ -0,0 +1,123 @@
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -o fakes/fake_registry_client.go ../../../../vendor/github.com/operator-framework/operator-registry/pkg/api.RegistryClient
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -o fakes/fake_registry_interface.go ../../registry/registry_client.go ClientInterface
+package catalog
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/operator-framework/operator-registry/pkg/api"
+	"github.com/operator-framework/operator-registry/pkg/client"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry"
+)
+
+const SkipPackageAnnotationKey = "olm.skipRange"
+
+type SourceRef struct {
+	Address     string
+	Client      client.Interface
+	LastConnect metav1.Time
+	LastHealthy metav1.Time
+}
+
+type SourceQuerier interface {
+	FindReplacement(currentVersion *semver.Version, bundleName, pkgName, channelName string, initialSource registry.CatalogKey) (*api.Bundle, *registry.CatalogKey, error)
+	Queryable() error
+}
+
+type NamespaceSourceQuerier struct {
+	sources map[registry.CatalogKey]registry.ClientInterface
+}
+
+var _ SourceQuerier = &NamespaceSourceQuerier{}
+
+func NewNamespaceSourceQuerier(sources map[registry.CatalogKey]registry.ClientInterface) *NamespaceSourceQuerier {
+	return &NamespaceSourceQuerier{
+		sources: sources,
+	}
+}
+
+func (q *NamespaceSourceQuerier) Queryable() error {
+	if len(q.sources) == 0 {
+		return fmt.Errorf("no catalog sources available")
+	}
+	return nil
+}
+
+func (q *NamespaceSourceQuerier) FindReplacement(currentVersion *semver.Version, bundleName, pkgName, channelName string, initialSource registry.CatalogKey) (*api.Bundle, *registry.CatalogKey, error) {
```

Comment:
I kinda wish we get rid of this as well. This func call is just to detect if there is a new version by making multiple API calls to registry to query the info. It doesn't really matter to the resolution which is run later if there is a change. Sometimes, `FindReplacement` will find something that will replace existing version and change the Subscriptions status to `UpgradeAvailable`. However, when the resolver runs, the replacement may not meet the constraints so no upgrade happens. That's a bit confusing. Plus, may as well let the resolver run because maybe it is less expensive (API-call-wise) that way. That part is unknown so it is fine to keep it as it is.

Reply:
I feel the same way, but I didn't want to take on that change right now.